### ISSUE TITLE
Ingestion Planning 1.37.0 1/2

### ISSUE THREAD

**Issue by yakkomajuri** [PostHog/posthog#9958]
<details>

<summary><b>Original content</b></summary>

## Context

We have 3 days left on the current sprint but we seem to be very much on track to wrap up both Yeetcode and the plugin server split (see https://github.com/PostHog/posthog/issues/9763). As a result, it's time to look ahead! How should we continue our work going forward?

Some useful frames of reference include:

- https://github.com/PostHog/posthog/issues/9652
- https://github.com/PostHog/posthog/issues/8957

**Please read the "Suggestion" section at the end.**

## Tasks

I'd still argue for working on plugin server reliability/maintainability for at least another sprint. Do push back if you disagree, but it seems the team is in agreement regarding this.

The one thing that could be worth considering outside of this scope is figuring out how to make the buffer work. Nevertheless, here's a superlist with suggested tasks (in no particular order):

### Things we agreed on before

We already said we would prioritize:

- #9856
- #9956

### Dead letter queue

- https://github.com/PostHog/posthog/issues/9189
- https://github.com/PostHog/posthog/issues/9780
- https://github.com/PostHog/posthog/issues/8266

### Plugin VMs

- https://github.com/PostHog/posthog/issues/7133
- https://github.com/PostHog/posthog/issues/8829
- With the above done, revisit and reenable historical exports on Cloud
- Plugins disabled silently

### Postgres queries

- https://github.com/PostHog/posthog/issues/8377
- https://github.com/PostHog/posthog/issues/8343
- https://github.com/PostHog/posthog/issues/8379

### Plugin server management

- Improve healthcheck
- Add readiness check
- Add plugin server graphs and alerts to self-hosted Grafana

### Persons on events blockers

- Figure out a way to make the buffer work without (significantly) impacting performance

### Miscellaneous

- https://github.com/PostHog/charts-clickhouse/pull/243
- Kafka connectivity issues (outlined [here](https://github.com/PostHog/posthog/issues/9652))
- Runbooks
- Plugins management activity log (who updated, disabled, when)
- https://github.com/PostHog/posthog/issues/8538

## Suggestion

I'd advocate for staying as close as possible to the theme of making the plugin server itself more robust before venturing out into other projects. As such, I'd punt on things like the dead letter queue, and work on fixing as much as possible of what's broken.

Thus, my tentative final list would be:

1. #9856
2. #9956
3. #8829 (and, paired with the server split, subsequently unblock releasing historical exports on Cloud again)
4. #8377
5. #8343
6. #8379
7. Kafkajs connectivity issues
8. Add readiness check, improve healthchecks across the board
9. https://github.com/PostHog/charts-clickhouse/pull/243

**Questions:**
- How do we group these into two measurable goals and split the work between the pairs?
- I feel like optimistically this is too little, so what should we subsequently work on? Dead letter queue work feels like a worthy project?

**Notes:**

- @macobo is support hero for the first week
- @Twixes is the release owner 

</details>

--------

### 1. Fix the main known issues regarding reliability and data integrity (owner: @tiina303)

@tiina303 @Twixes 

**Why:** To make the plugin server more reliable and prevent churn from data integrity issues
**Why now:** To solve pressing customer issues and make us comfortable deploying large changes to ingestion

We aim to complete all of the following:

- [ ] https://github.com/PostHog/posthog/issues/9956
- [x] https://github.com/PostHog/posthog/issues/8829 (and, paired with the server split, subsequently unblock releasing historical exports on Cloud again)
- [ ] Add metric for persons/groups out of sync between Postgres <> CH 
- [ ] https://github.com/PostHog/posthog/issues/8343
- [x] Investigate kafkajs connectivity issues
- [ ] https://github.com/PostHog/charts-clickhouse/pull/243 

### 2. Complete the final ingestion-related tasks for persons/groups on events (owner: @macobo)

@macobo @yakkomajuri 

**Why:** To unlock high scale querying
**Why now:** To ensure our ingestion pipeline guarantees hold true in its new proposed state

We aim to complete all of the following:

- [ ] Get the buffer live
- [x] Consume events in order
- [ ] Add `created_at` for persons and groups as a column 
- [x] Add a metric for E2E ingestion time

Our stretch goals are:

- Write up the async migration to backfill data (both persons/groups as well as using JSON data type)
- Ship live events: if we can collaborate with App West

**Comment by tiina303** [comment]
Regarding grouping From what you wrote 2 & postgres stuff belong together, 1 is somewhat related too. Furthermore I'd say that this is potentially quite difficult to fix, so not too worried about it being too little work.

**Comment by macobo** [comment]
Notes on proposed tasks:
- https://github.com/PostHog/posthog/issues/9856 is something we'll ship already as part of this sprint - need to write some backend tests still though.
- Not sure what "Kafkajs connectivity issues" entails even after reading up on it. I guess I'd need to live through at least one of these troubles. :sweat:
- I don't think https://github.com/PostHog/posthog/issues/8379 is relevant for us - cohorts aren't processed via plugin-server and @neilkakkar / @EDsCODE have been working to change this logic completely. We can verify though and potentially close the issue.
- https://github.com/PostHog/posthog/issues/8377 not sure if this belongs to this team either.
- Plugin retries question: Would it make sense to prioritize persons-on-events over this? This is a new feature for our end-users and I think it might make sense to work on this together with e.g. exposing person_id to plugins later down the line?

Meta question: Can we define a reliability metric for ingestion? E.g. no more than X% of events fail and check whether we're hitting that after these tasks or not?

Suggestions:
- Let's get the buffer live. With the refactoring we've done splitting the plugin-server we should be mostly ready, what remains is debugging FUD/remaining issues.
- Given we are going to be tackling some data consistency problems, let's add a cronjob emitting a metric for how many persons have properties out of sync with postgresql. Same for groups.
- Ambitious: Let's kick off backfilling persons-on-events on cloud.

**Comment by yakkomajuri** [comment]
Thanks a lot for this @macobo! The Postgres errors particularly I just grabbed from a group somewhere else without looking into them - thanks for the digging.

Some clarifications:

- **Kafkajs connectivity:** It seems if one of our _n_ brokers goes down, kafkajs crashes and ingestion stops. Currently we have a hacky fix via the healthcheck, which will just take the task down and then when reconnecting things work again, but this is something we need to dig into and maybe even look upstream for answers/solutions.
- **Plugin retries:** I'd argue against that. This seems like "it's just a user feature", _however_, it's something that's broken and has bit us in the past. Currently we have 2 things going on: historical exports are disabled on Cloud (although I refactored some things and believe they should be safe), and retries have been manually reduced. I'd like us to work on this so we can determine what guidelines to live by, and bring back an important feature (historical exports). Also note that this work might include some reliability work around plugins ending up disabled in some but not all instances, and refactors/fixes like https://github.com/PostHog/posthog/pull/9943 of the LazyVM that have caused customer problems.

**New suggestions:**

- https://github.com/PostHog/posthog/issues/9956
- https://github.com/PostHog/posthog/issues/8829 (and, paired with the server split, subsequently unblock releasing historical exports on Cloud again)
- Get the buffer live: this makes sense to do alongside this entire ingestion refactor as the buffer will be a key part going forward
- Key metrics: Postgres <> CH sync, Ingestion E2E time
- https://github.com/PostHog/posthog/issues/8343
- Kafkajs connectivity issues
- Add readiness check, improve healthchecks across the board
- https://github.com/PostHog/charts-clickhouse/pull/243 
- Consuming events in order
- Live events: adding this as the last item as it is a key part of shipping the buffer and we might collaborate with @alexkim205 here

**Comment by yakkomajuri** [comment]
## WIP - Notes from planning

Retries:

- Define clear goals around retries

Secondary goal:

- Re-enable historical exports on Cloud 
    - Add a metric for exportEvents delivery rate and errors (set an internal target to monitor e.g. 99%)

- Buffer > live events

- Yeet benchmarks (but make sure we measure e2e ingestion on Cloud!)

- Async migration for the persons on events & JSON column backfill should be a priority for the next sprint

----------

- Fix all known data integrity issues
- Get the ingestion pipeline to its "final form" as currently planned

Buckets:

- Complete these tasks on data integrity and reliability `@tiina303`
- Complete these tasks on persons on events scaffolding `@macobo`

**Comment by tiina303** [comment]
Suggestion for changes of work assignments:

@tiina303  @Twixes 

- [ ] https://github.com/PostHog/posthog/issues/9181  @tiina303 
     i.e. consume events in order currently under Yakko/Karl
     this will fix the main causes of:
        - https://github.com/PostHog/posthog/issues/9956
        - https://github.com/PostHog/posthog/issues/8343
        - https://github.com/PostHog/posthog/issues/10058
- [ ] https://github.com/PostHog/posthog/issues/8829 @Twixes 
         

@yakkomajuri  @macobo 

- [ ] everything kafka (as you have been looking into this already it doesn't make sense to switch)
      - Investigate kafkajs connectivity issues
      - https://github.com/PostHog/charts-clickhouse/pull/243


TBD (I propose Michael and I help out here if Kafka is still on 🔥) 
- [ ] Get the buffer live
- [ ] Add `created_at` for persons and groups as a column
- [ ] Add a metric for E2E ingestion time

Scrap
- [x] Add metric for persons/groups out of sync between Postgres <> CH
     Created an issue for the problem: https://github.com/PostHog/posthog/issues/10058 ; ordered processing of events will resolve or make it rare. Furthermore it will not be relevant once we have persons on events adding metrics here shouldn't be a priority.

**Comment by yakkomajuri** [comment]
Disagree with scrapping the metric - how do you know you fixed the issue if you don't have a metric? Also note that as a general rule trying to keep 2 systems in sync is impossible - there will always be drift. It's worth measuring this anyway

**Comment by tiina303** [comment]
> Furthermore it will not be relevant once we have persons on events

That was the main reason I proposed scrapping the metrics goal. I assume we'll drop the person table in CH (as like you say 2 systems in sync is impossible) & look it up from Postgres again for the persons page.

**Comment by macobo** [comment]
persons-on-events will not drop the persons table - it just removes it from _insights_ queries

This metric is and will be relevant.

**Comment by neilkakkar** [comment]
Yes^. For more (helpful?) context:

Some queries (like persons cohorts) still need the persons table, since it's only querying persons, not events.

And querying persons from postgres is too slow, we want to move that to clickhouse persons table too: https://github.com/PostHog/posthog/issues/8660

**Comment by tiina303** [comment]
ok, thanks for the context, so that will still be under me & Michael then, the rest of my proposal stands.
